File: src/main/java/mods/betterfoliage/client/misc/BlockMatcher.java
package mods.betterfoliage.client.misc;

import java.util.Collection;
import java.util.Iterator;
import java.util.Set;

import mods.betterfoliage.client.util.ResourceUtils;
import net.minecraft.block.Block;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.event.world.WorldEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import com.google.common.collect.Sets;

@SideOnly(Side.CLIENT)
public class BlockMatcher {

	public Set<Class<?>> whiteList = Sets.newHashSet();
	public Set<Class<?>> blackList = Sets.newHashSet();
	public Set<Integer> blockIDs = Sets.newHashSet();

	public boolean matchesClass(Block block) {
		for (Class<?> clazz : blackList) if (clazz.isAssignableFrom(block.getClass())) return false;
		for (Class<?> clazz : whiteList) if (clazz.isAssignableFrom(block.getClass())) return true;
		return false;
	}
	
	public boolean matchesID(int blockId) {
		return blockIDs.contains(blockId);
	}
	
	public boolean matchesID(Block block) {
		return blockIDs.contains(Block.blockRegistry.getIDForObject(block));
	}
	
	public void updateClassLists(String[] newWhitelist, String[] newBlacklist) {
	    whiteList.clear();
	    for(String className : newWhitelist) try {
	        whiteList.add(Class.forName(className));
	    } catch(ClassNotFoundException e) {}
	    
	    blackList.clear();
        for(String className : newBlacklist) try {
            blackList.add(Class.forName(className));
        } catch(ClassNotFoundException e) {}
        
        updateBlockIDs();
	}
	
	@SuppressWarnings("unchecked")
    public void updateBlockIDs() {
	    blockIDs.clear();
        Iterator<Block> iter = Block.blockRegistry.iterator();
        while (iter.hasNext()) {
            Block block = iter.next();
            if (matchesClass(block)) blockIDs.add(Block.blockRegistry.getIDForObject(block));
        }
	}
	
	public static void loadDefaultLists(ResourceLocation defaults, Collection<String> blacklist, Collection<String> whitelist) {
	    for (String line : ResourceUtils.getLines(defaults)) {
	        if (line.startsWith("-"))
                blacklist.add(line.substring(1));
            else
                whitelist.add(line);
	    }
	}
	
	/** Caches block IDs on world load for fast lookup
	 * @param event
	 */
	@SubscribeEvent
	public void handleWorldLoad(WorldEvent.Load event) {
		updateBlockIDs();
	}
}

File: src/main/kotlin/mods/betterfoliage/client/config/BlockMatcher.kt
package mods.betterfoliage.client.config

import mods.octarinecore.client.gui.NonVerboseArrayEntry
import mods.octarinecore.client.resource.get
import mods.octarinecore.client.resource.getLines
import mods.octarinecore.client.resource.resourceManager
import mods.octarinecore.common.config.ConfigPropertyBase
import mods.octarinecore.metaprog.getJavaClass
import net.minecraft.block.Block
import net.minecraft.client.multiplayer.WorldClient
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.common.config.Configuration
import net.minecraftforge.common.config.Property
import net.minecraftforge.event.world.WorldEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

/**
 * Match blocks based on their class names. Caches block IDs for faster lookup.
 *
 * @param[domain] resource domain for defaults file
 * @param[path] resource path for defaults file
 */
class BlockMatcher(val domain: String, val path: String) : ConfigPropertyBase() {

    val blackList = mutableListOf<Class<*>>()
    val whiteList = mutableListOf<Class<*>>()
    val blockIDs = hashSetOf<Int>()
    var blacklistProperty: Property? = null
    var whitelistProperty: Property? = null

    fun matchesClass(block: Block): Boolean {
        val blockClass = block.javaClass
        blackList.forEach { if (it.isAssignableFrom(blockClass)) return false }
        whiteList.forEach { if (it.isAssignableFrom(blockClass)) return true }
        return false
    }
    fun matchesID(block: Block) = blockIDs.contains(Block.blockRegistry.getIDForObject(block))
    fun matchesID(blockId: Int) = blockIDs.contains(blockId)

    override fun attach(target: Configuration, langPrefix: String, categoryName: String, propertyName: String) {
        lang = null
        val defaults = readDefaults(domain, path)
        blacklistProperty = target.get(categoryName, "${propertyName}Blacklist", defaults.first)
        whitelistProperty = target.get(categoryName, "${propertyName}Whitelist", defaults.second)
        listOf(blacklistProperty!!, whitelistProperty!!).forEach {
            it.setConfigEntryClass(NonVerboseArrayEntry::class.java)
            it.setLanguageKey("$langPrefix.$categoryName.${it.name}")
        }
        read()
    }

    override fun read() {
        listOf(Pair(blackList, blacklistProperty!!), Pair(whiteList, whitelistProperty!!)).forEach {
            it.first.clear()
            it.first.addAll(it.second.stringList.map { getJavaClass(it) }.filterNotNull())
        }
        updateIDs()
    }

    fun updateIDs() {
        blockIDs.clear()
        Block.blockRegistry.forEach {
            if (matchesClass(it as Block)) blockIDs.add(Block.blockRegistry.getIDForObject(it))
        }
    }

    override val hasChanged: Boolean
        get() = blacklistProperty?.hasChanged() ?: false || whitelistProperty?.hasChanged() ?: false

    override val guiProperties: List<Property> get() = listOf(whitelistProperty!!, blacklistProperty!!)

    fun readDefaults(domain: String, path: String): Pair<Array<String>, Array<String>> {
        val blackList = arrayListOf<String>()
        val whiteList = arrayListOf<String>()
        val defaults = resourceManager[domain, path]?.getLines()
        defaults?.map{ it.trim() }?.filter { !it.startsWith("//") && it.isNotEmpty() }?.forEach {
            if (it.startsWith("-")) { blackList.add(it.substring(1)) }
            else { whiteList.add(it) }
        }
        return (blackList.toTypedArray() to whiteList.toTypedArray())
    }

    @SubscribeEvent
    fun onWorldLoad(event: WorldEvent.Load) { if (event.world is WorldClient) updateIDs() }

    init { MinecraftForge.EVENT_BUS.register(this) }

}